sharedbuffers/mapped_struct.py

```diff
@@ -1719,6 +1737,374 @@ def __str__(self):
+            return cython.cast(proxied_frozenset, self)._frozenset_eq(seq)
+        else:
+            return cython.cast(proxied_frozenset, seq)._frozenset_eq(self)
+
+    @cython.ccall
+    @cython.locals(strict_subset=cython.bint, i=cython.Py_ssize_t,
+        xlen=cython.Py_ssize_t, pfset='proxied_frozenset',
+        seqlen=cython.Py_ssize_t, bitrep=cython.size_t, dcode=cython.char,
+        objlen=cython.longlong, itemsize=cython.uchar, offset=cython.Py_ssize_t,
+        dcode2=cython.char, oblen2=cython.longlong, itemsize2=cython.uchar,
+        offset2=cython.Py_ssize_t, h1=cython.ulonglong, h2=cython.ulonglong)
+    def _subset(self, seq, strict_subset):
+        i = 0
+        xlen = len(self)
+
+        if isinstance(seq, proxied_frozenset):
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                if self.objlist is None:
+                    # Both proxies use compressed representation.
+                    return (self.bitrep_lo & pfset.bitrep_lo) == self.bitrep_lo and (
+                        self.bitrep_hi & pfset.bitrep_hi) == self.bitrep_hi and (
+                        not strict_subset or self.bitlen < pfset.bitlen)
+                else:
+                    dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+                    for i in xrange(xlen):
+                        if self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None) not in pfset:
+                            return False
+                    return not strict_subset or xlen < pfset.bitlen
+
+            elif self.objlist is None:
+                bitrep = self.bitrep_lo
+                while bitrep != 0:
+                    if (bitrep & 1) != 0 and i not in pfset:
+                        return False
+                    bitrep >>= 1
+                    i += 1
+
+                bitrep = self.bitrep_hi
+                while bitrep != 0:
+                    if (bitrep & 1) != 0 and i not in pfset:
+                        return False
+                    bitrep >>= 1
+                    i += 1
+
+                return not strict_subset or xlen < len(pfset)
+
+            # Both proxies use sorted lists.
+            j = 0
+            seqlen = len(pfset.objlist)
+            if xlen > seqlen:
+                return False
+
+            dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+            if dcode in ('q', 'I', 'i', 'd') and cython.compiled:
+                # fast path, use the search_hkey variants
+                dcode2, objlen2, itemsize2, offset2, _struct2 = pfset.objlist._metadata()
+                for i in xrange(xlen):
+                    val = self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None)
+                    j = pfset._search_key(val, dcode, offset, j, seqlen - j, j)
+                    if pfset.objlist._c_getitem(j, dcode2, objlen2, itemsize2, offset2, _struct2, None) != val:
```

Comment:
If search_hkey fails to find a key, it may retun an index beyond the end. Then getitem would raise an IndexError. Beware of that, you have to verify the returned index.